### ISSUE TITLE
Trailing stoploss on sell_signal

### ISSUE THREAD

**Issue by devilsinthedetail** [freqtrade/freqtrade#4271]
As the title suggests, it would be nice to have a trailing stoploss activated on sell_signal. I searched for other issues but could not find something like that and not sure if this easy to implement with [custom stoplosses](https://www.freqtrade.io/en/latest/strategy-advanced/#time-based-trailing-stop).

So it would be nice to have the trailing stoploss to be adjusted when sell_signal is coming in, making the stoploss the current value (where it would sell based on sell_signal) but in this case it would just use the trailing stoploss to sell. 

Example with Trailing Stop at lets say 2%. 

**How it currently works:**
- Trade 1 Buy at 100. 
- (Trade goes to 102, trailing stoploss is activated and trade is closed at some point)
- Trade goes to 80
- Trade goes to 85 (sell_signal is triggered and trade is closed)

**How it would be nice to work:**
- Trade 1 Buy at 100. 
- (Trade goes to 102, trailing stoploss is activated and trade is closed at some point)
- Trade goes to 80
- Trade goes to 85 (trailing stoploss is adjusted down from 2% to -15% and trails from there and is closed at some point)

**Comment by xmatthias** [comment]
With enough creativity (and by reading the docs), it's easily possible with custom stoplosses.

**Comment by devilsinthedetail** [comment]
Thanks for your reply, I may lack the creativity I guess. The custom stopploss has these parameters: 
```
        :param pair: Pair that's about to be sold.
        :param trade: trade object.
        :param rate: Rate that's going to be used when using limit orders
        :param **kwargs: Ensure to keep this here so updates to this won't break your strategy.
        :return float: New stoploss value, relative to the open price
```
I don't really see how I can know if my sell_signal is triggered from these? Or I guess I don't understand it fully. I just thought i I can somehow reset
```
    trailing_stop_positive = 0.02
    trailing_stop_positive_offset = 0.03
```
these two when a sell_signal comes in?

**Comment by xmatthias** [comment]
with Creativity, i mean the idea to combine different methods (like [get_analyzed_dataframe()](https://www.freqtrade.io/en/latest/strategy-customization/#get_analyzed_dataframepair-timeframe)).
It's available in live / dry run modes only, but will give you what you need.

However: you CANNOT assign the sell-signal to 'sell' in this case, as then it'd immediately trigger (sell-signal does not really care about what you think you do with stoplosses) - if a sell-signal is there (and no buy signal) - it'll sell.

Instead, create a "0" column for sell `dataframe.loc[:, 'sell'] = 0` - and assign it to `sell_mine` (or whatever) and use that to determine the offset.

**Comment by devilsinthedetail** [comment]
Okay I see. I'm still not 100% sure how to code it. Is something like that correct:

```python
def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs) -> float:

    if self.dp:
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=metadata['pair'],
                                                                timeframe=self.timeframe)
    if dataframe['sell_mine']:
        desired_stoploss = current_profit # ???
        return max(min(desired_stoploss, 0.05), 0.025) # ???
    else:
        return -1
```
I took this from the examples but I don't really get how this (max(min(desired_stoploss, 0.05), 0.025)) correlates to the two trailing values?
```
    trailing_stop_positive = 0.02
    trailing_stop_positive_offset = 0.03
```
trailing_stop_positive_offset would need to be equal to current_profit right? 

Edit:
Maybe this makes more sense:
```python
def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs) -> float:

    if self.dp:
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=metadata['pair'],
                                                                timeframe=self.timeframe)
    if dataframe['sell_mine'] and not dataframe['buy']:
        return (current_profit - self.trailing_stop_positive)
    else:
        return -1
```

**Comment by xmatthias** [comment]
> max(min(desired_stoploss, 0.05), 0.025)

you should read the explanation in the docs for this.
It only makes sense in context of the full sample - where `desired_stoploss` is calculated based on current profit - and this min/max is creating a channel of trailing with a minimum of 2.5%, and a maximum of 5%.


Your edited thing makes more sense for your particular case - but you'll need to check the last row (and you cannot really backtest this as is, as neither get_analyzed_dataframe, nor "last row" are available in backtesting).

**Comment by devilsinthedetail** [comment]
Cool thanks, its a pity that I can't backtest it but that gives less freedom to overfit on historical data, so thats good. Do you think this is correct so far? Then I would just try it in dry mode. 
```python
def populate_sell_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
    dataframe.loc[:, 'sell'] = 0 # In order for traditional logic to work as expected
    dataframe.loc[:, 'sell_mine'] = 0 # Make sure its not comparing against NaN
    dataframe.loc[my_sell_condition, 'sell_mine'] = 1
    return dataframe

def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs) -> float:
    if self.dp:
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=metadata['pair'],
                                                                timeframe=self.timeframe)
    if dataframe['sell_mine'].iloc[-1] and not dataframe['buy'].iloc[-1]:
        return (current_profit - self.trailing_stop_positive)
    else:
        return -1
```

**Comment by xmatthias** [comment]
it seems okish as far as i can tell (the best test is always dry-run, where you see how it's moving).

depending on your sell_mine signal, it will NOT be trailing though - but it'll adjust at every `sell_mine` - but between these signals it'll stay stuck - so assuming the pair raises further, your stoploss will remain in place.

**Comment by devilsinthedetail** [comment]
Oh I see, because it might alternate between 1 and 0. So how about I (ab)use the custom stoploss for something like that:
```python    
def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs) -> float:
    original_trailing = 0.02
    if self.dp:
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=metadata['pair'],
                                                                timeframe=self.timeframe)
    if dataframe['sell_mine'].iloc[-1] and not dataframe['buy'].iloc[-1]:
        self.trailing_stop_positive_offset = current_profit
    else:
        self.trailing_stop_positive_offset = original_trailing
    return -1
```

Edit: Actually it should never be reset to the original trailing loss:
```python    
def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs) -> float:
    if self.dp:
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=metadata['pair'],
                                                                timeframe=self.timeframe)
    if dataframe['sell_mine'].iloc[-1] and not dataframe['buy'].iloc[-1]:
        self.trailing_stop_positive_offset = current_profit
    return -1
```
Edit 2: No sorry, it has to be reset for the next trades:
```python    
def custom_stoploss(self, pair: str, trade: Trade, current_time: datetime, current_rate: float,
                    current_profit: float, **kwargs) -> float:
    original_trailing = 0.02
    if self.dp:
        dataframe, last_updated = self.dp.get_analyzed_dataframe(pair=metadata['pair'],
                                                                timeframe=self.timeframe)
    if dataframe['sell_mine'].iloc[-1] and not dataframe['buy'].iloc[-1]:
        self.trailing_stop_positive_offset = current_profit
    elif dataframe['buy'].iloc[-1]:
        self.trailing_stop_positive_offset = original_trailing
    return -1
```

**Comment by xmatthias** [comment]
what counts here is what you return - not what you assign, so i doubt this will work

you might be better off using the trade object to determine if you're in a "trailing" scenario or not ...

**Comment by devilsinthedetail** [comment]
Okay I tried it a bit in dry run mode and both versions have problems. I guess setting this:
`self.trailing_stop_positive_offset = current_profit`
will change the strategy for all open trades. Is that correct? Meaning when one sell signal is coming in, it will potentially close all open trades? Thats at least what I got from the 3 sell signals I got in this short timeperiod. 

Also, how would I determine if I'm in a trailing scenario from the Trades object? Thanks! 

Edit: Just looking at the code for models and interface. I think I might figure something out myself. Thanks anyway.

**Comment by xmatthias** [comment]
you should really read the documentation for custom stoploss.
it includes multiple samples - none of which will asign anything to the `self.trailing_* ` properties.
By doing that, the behaviour will become undefined.

Instead, you'll return the the relative distance the new stoploss should take from the current profit.
